### PR TITLE
Conservative search for target epsilon in get_noise_multiplier

### PR DESCRIPTION
**Pull request by alexandresablayrolles**
## Problem

@tholop noticed in #346 that `get_noise_multiplier` can overshoot the target epsilon when searching for highest possible noise multiplier. This is because of the binary search implementation.

## Changes

- Noise multiplier now verifies that the actual epsilon is *lower* than the target epsilon.
  - Added test for this.  
- Bounds of the range of search for sigma, and precision can now be given as arguments to `get_noise_multiplier`.

Differential Revision: D34003401

### CODE REVIEW

File: opacus/accountants/utils.py [hunk 1]
```diff
@@ -39,27 +38,31 @@ def get_noise_multiplier(
...
 
     while eps > target_epsilon:
-        sigma_max = 2 * sigma_max
-        accountant.steps = [(sigma_max, sample_rate, int(epochs / sample_rate))]
+        sigma_high = 2 * sigma_high
+        accountant.steps = [(sigma_high, sample_rate, steps)]
         eps = accountant.get_epsilon(delta=target_delta, **kwargs)
-        if sigma_max > MAX_SIGMA:
+        if sigma_high > sigma_max:
             raise ValueError("The privacy budget is too low.")
```
Comment (romovpa) [inline opacus/accountants/utils.py:56]: Here we try to automatically initialize `sigma_high` for the binary search. This behaviour is not consistent with the docstring: I can set a specific bound but it can change it anyway. Also, having both `sigma_high` and `sigma_max` is counter-intuitive and it takes knowing the whole algorithm to understand what is the difference.

## Proposal

Leave only `sigma_low`, `sigma_high` in parameters. I would also rename to min/max to emphasize that the result will be bounded:

```python
def get_noise_multiplier(
    ...
    sigma_min: float = 0.01,
    sigma_max: Optional[float] = None,
)
```

If `sigma_max is None`, run this part to get the upper bound starting from `sigma_max+1`. 

Leave MAX_SIGMA a global constant, maybe make it higher.

File: opacus/accountants/utils.py [hunk 2]
```diff
@@ -1,24 +1,18 @@
...
     target_delta: float,
     sample_rate: float,
-    epochs: int,
+    epochs: int = None,
+    steps: int = None,
     accountant: str = "rdp",
+    sigma_precision=0.01,
+    sigma_low=0.01,
+    sigma_high=10,
+    sigma_max=2000,
```
Comment (romovpa) [inline opacus/accountants/utils.py:15]: Add type definitions please

File: opacus/accountants/utils.py [hunk 3]
```diff
@@ -1,24 +1,18 @@
...
 def get_noise_multiplier(
     *,
     target_epsilon: float,
     target_delta: float,
     sample_rate: float,
-    epochs: int,
+    epochs: int = None,
+    steps: int = None,
     accountant: str = "rdp",
+    sigma_precision=0.01,
```
Comment (romovpa) [inline opacus/accountants/utils.py:12]: Let's rename it to `sigma_tolerance`, which is the common word used in optimization (see [scipy.bisect](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.bisect.html)).

Isn't this precision too low? I mean we should make this value lower, closer to typical float precision. Btw, this higher/lower confusion is another reason to rename the argument to tolerance.

File: opacus/accountants/utils.py [hunk 4]
```diff
@@ -1,24 +1,18 @@
...
-
-
 def get_noise_multiplier(
     *,
     target_epsilon: float,
     target_delta: float,
     sample_rate: float,
-    epochs: int,
+    epochs: int = None,
+    steps: int = None,
```
Comment (romovpa) [inline opacus/accountants/utils.py:10]: If the argument can be None, it should have type `Optional[int]`

File: opacus/accountants/utils.py [hunk 5]
```diff
@@ -39,27 +38,31 @@ def get_noise_multiplier(
...
 
         if eps < target_epsilon:
-            sigma_max = sigma
+            sigma_high = sigma
         else:
-            sigma_min = sigma
+            sigma_low = sigma
 
-    return sigma
+    return sigma_high
```
Comment (romovpa) [inline opacus/accountants/utils.py:68]: A possible corner case when `sigma_high < sigma_low` is not considered here.

File: opacus/accountants/utils.py [hunk 6]
```diff
@@ -39,27 +38,31 @@ def get_noise_multiplier(
...
+        sigma = (sigma_low + sigma_high) / 2
+        accountant.steps = [(sigma, sample_rate, steps)]
         eps = accountant.get_epsilon(delta=target_delta, **kwargs)
 
         if eps < target_epsilon:
-            sigma_max = sigma
+            sigma_high = sigma
         else:
-            sigma_min = sigma
+            sigma_low = sigma
```
Comment (romovpa) [inline opacus/accountants/utils.py:66]: I'm thinking if we can use [`scipy.optimzie.bisect()`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.bisect.html) instead of implementing our own.
It is easy to implement binary search incorrectly so we need to make tests for all corner cases. If we use scipy, we don't need to add our own tests, plus it is much more stable than the rest of our dependencies.

File: opacus/accountants/utils.py (hunk 3 above)
Comment (romovpa) [inline opacus/accountants/utils.py:12]: You need to add check that `sigma_tolerance` is strictly positive. Alternative: pass is to scipy.bisect and let it do the check for you.

**Review by romovpa — CHANGES_REQUESTED** [review]

File: opacus/accountants/utils.py [hunk 7]
```diff
@@ -53,27 +51,40 @@ def get_noise_multiplier(
...
+    if eps_high > target_epsilon:
+        raise ValueError(
+            f"Noise multiplier of {sigma_high} is not enough to have a privacy budget of {target_epsilon}, please use a higher sigma_high."
+        )
+
+    accountant.steps = [(sigma_low, sample_rate, steps)]
+    eps = accountant.get_epsilon(delta=target_delta, **kwargs)
+    if eps < target_epsilon:
+        warnings.warn(
+            f"Noise multiplier of {sigma_low} is still too high for a privacy budget of {target_epsilon}, please use a lower sigma_low."
```
Comment (romovpa) [inline opacus/accountants/utils.py:75]: I would replace "please use a lower sigma_low" with something like:
"you can benefit quality by allowing smaller sigma_low without exceeding target epsilon"

But the current version is also fine.

File: opacus/accountants/utils.py [hunk 8]
```diff
@@ -13,26 +13,20 @@
...
     *,
     target_epsilon: float,
     target_delta: float,
     sample_rate: float,
-    epochs: int,
+    epochs: Optional[int] = None,
+    steps: Optional[int] = None,
     accountant: str = "rdp",
+    epsilon_tolerance: float = 0.01,
+    sigma_low: float = 0.01,
```
Comment (romovpa) [inline opacus/accountants/utils.py:28]: I wanted to ask, why the default `sigma_low` is so high. We will surprise the user with the warning and suboptimal result if the best sigma is 1e-3 for example.
But then I found that I don't understand what is the point of having this parameter at all? Why not starting with sigma_low=0 always? 

This won't hurt this expression `target_epsilon - eps_high > epsilon_tolerance`. We will never reach this value closer than tolerance/2.

File: opacus/accountants/utils.py [hunk 9]
```diff
@@ -13,26 +13,20 @@
...
+    sigma_high: float = 200,
```
Comment (romovpa) [inline opacus/accountants/utils.py:29]: Why we are so sure that 200 is high enough? 

Why not eliminate this constant by probing `sigma_high = tol*2^k` using this removed algorithm:
```
    sigma_high = tol
    while eps > target_epsilon:
        sigma_high = 2 * sigma_high
        accountant.steps = [(sigma_high, sample_rate, int(epochs / sample_rate))]
        eps = accountant.get_epsilon(delta=target_delta, **kwargs)
        if sigma_high > MAX_SIGMA:
            raise ValueError("The privacy budget is too low.")
```

**Review by romovpa — CHANGES_REQUESTED** [review]

File: opacus/accountants/utils.py [hunk 10]
```diff
@@ -13,26 +13,19 @@
...
-DEFAULT_SIGMA_MIN_BOUND = 0.01
-# starting point for a max range bound when searching for sigma given epsilon
-DEFAULT_SIGMA_MAX_BOUND = 10
-# condition to halt binary search for sigma given epsilon
-SIGMA_PRECISION = 0.01
-# max possible value for returned sigma.
-# Noise higher than MAX_SIGMA considered unreasonable
-MAX_SIGMA = 2000
-
+MAX_SIGMA = 1e6
```
Comment (romovpa) [inline opacus/accountants/utils.py:18]: There was a comment before which I recommend to leave because the constant looks a bit surprising.

```python
# Noise level higher than MAX_SIGMA considered unreasonable 
# leading to "not sufficient privacy budget" exception
MAX_SIGMA = 1e6
```

**Review by romovpa — APPROVED** [review]
Sweet! 
One nit to add a comment.
